code/network/multi_interpolate.h

```diff
@@ -0,0 +1,127 @@
+	// we already received a newer packet than this one for that type of info.
+	int _hull_comparison_frame;						// what frame was the last hull information received?
+	int _shields_comparison_frame;					// what frame was the last shield information received?
+	SCP_vector<int> _subsystems_comparison_frame;	// what frame was the last subsystem information received? (for each subsystem)
+	int _ai_comparison_frame;						// what frame was the last ai information received?
+
+public:
+
+	// adds a new packet, whilst also manually sorting the relevant entries
+	void add_packet(int frame, int time_delta, vec3d* position, vec3d* velocity, vec3d* rotational_velocity, vec3d* desired_velocity, vec3d* desired_rotational_velocity, angles* angles, int player_index);
+	void interpolate(vec3d* pos, matrix* ori, physics_info* pip, bool player_ship);
+
+	int get_hull_comparison_frame() { return _hull_comparison_frame; }
+	int get_shields_comparison_frame() { return _shields_comparison_frame; }
+	
+	int get_subsystem_comparison_frame(int i) 
+	{ 
+		if (i < static_cast<int>(_subsystems_comparison_frame.size()) && i >= 0) {
+			return _subsystems_comparison_frame[i]; 
+		} // if it somehow got passed nonsense, chances are what it is is trying to read is nonsense, and INT_MAX will keep it from doing anything crazy.
+		else {
+			return INT_MAX;
+		}
+	}
+
+	int get_ai_comparison_frame() { return _ai_comparison_frame; }
+
+	void set_hull_comparison_frame(int frame) { _hull_comparison_frame = frame; }
+	void set_shields_comparison_frame(int frame) { _shields_comparison_frame = frame; }
+
+	void set_subsystem_comparison_frames(int i, int frame)
+	{ 
+		if (i < static_cast<int>(_subsystems_comparison_frame.size()) && i >= 0) {
+			_subsystems_comparison_frame[i] = frame;
+		} 
+	}
+
+	void set_ai_comparison_frame(int frame) { _ai_comparison_frame = frame; }
+
+	void force_interpolation_mode() { _simulation_mode = true; }
+
+	void reset() 
+	{
+		_packets.clear();
+		_packets.reserve(PACKET_INFO_LIMIT * 2);
+		_upcoming_packet_index = -1;
+		_prev_packet_index = -1;
+		_simulation_mode = false;
+		_packets_expended = false;
+		_local_skip_forward = 0;
+		_hull_comparison_frame = -1;
+		_shields_comparison_frame = -1;
+		_source_player_index = -1;
+
+		// if we are resetting, that means that the ship is going to respawn, so don't clear it out, just set the values to invalid.
+		for (auto& frame : _subsystems_comparison_frame) {
+			frame = -1;
+		}
+
+		_ai_comparison_frame = -1;
+	}
+
+	interpolation_manager()
+	{
+		_packets.reserve(PACKET_INFO_LIMIT * 2);
+		_upcoming_packet_index = -1;
+		_prev_packet_index = -1;
+		_simulation_mode = false;
+		_packets_expended = false;
+		_local_skip_forward = 0;
+		_hull_comparison_frame = -1;
+		_shields_comparison_frame = -1;
+		_subsystems_comparison_frame.reserve(64);
+		_ai_comparison_frame = -1;
+		_source_player_index = -1;
+	}
```

Comment:
Since this is attached to general `object`'s, no memory should be allocated in this constructor. `Objects[]` is still a static size much of this allocated memory could end up wasted. And this is 100% wasted in single player too. Just have the `reset()` take care of it.